internal/entitycache/cache.go

```diff
@@ -0,0 +1,669 @@
+	for _, id := range iter.baseEntityBatch {
+		delete(baseEntities.entities, *id)
+	}
+	baseEntities.changed.Broadcast()
+}
+
+// addEntity adds an entity that has been received
+// from the underlying iterator. It reports whether the
+// entity has been added to iter.entityBatch.
+//
+// Called from iter.run without any locks held.
+func (iter *Iter) addEntity(e *mongodoc.Entity) bool {
+	entities := &iter.cache.entities
+	entities.mu.Lock()
+	defer entities.mu.Unlock()
+	if _, ok := entities.entities[*e.CanonicalURL()]; ok {
+		// The entity has already been fetched, or is being fetched.
+		// This also implies that its base entity has already been added (or
+		// is in the process of being added) to the cache.
+		return false
+	}
+	if entities.version == iter.version {
+		// The entity we have here is valid to put into the cache, so do that.
+		// Note: we know from the check above that the entity is not
+		// already present in the cache.
+		entities.addEntity(e, nil)
+	}
+	iter.entityBatch = append(iter.entityBatch, e)
+
+	// TODO should we use a single mutex for both stashes?
+	baseEntities := &iter.cache.baseEntities
+	baseEntities.mu.Lock()
+	defer baseEntities.mu.Unlock()
+
+	baseURL := baseURL(e.URL)
+	if _, ok := baseEntities.entities[*baseURL]; !ok {
+		// We need to fetch the base entity, so add it to our
+		// batch and signal that it will be fetched by adding it
+		// to the map. Note: this assumes that the client doing
+		// the iteration will make progress - it could delay
+		// other base entity reads arbitrarily by not calling
+		// Next. This should not be a problem in practice.
+		iter.baseEntityBatch = append(iter.baseEntityBatch, baseURL)
+		baseEntities.entities[*baseURL] = nil
+	}
+	return true
+}
+
+// sendBatch obtains all the batched base entities and sends all the
+// batched entities on iter.entityc. If it encounters an error, or the
+// iterator is closed, it sets iter.err and returns false.
+//
+// Called from iter.run with no locks held.
+func (iter *Iter) sendBatch() bool {
+	// Start a fetch for all base entities.
+	// TODO use actual batch fetch with $in etc rather
+	// than starting a goroutine for each base entity.
+	baseEntities := &iter.cache.baseEntities
+	baseEntities.mu.Lock()
+	iter.cache.wg.Add(len(iter.baseEntityBatch))
+	for _, id := range iter.baseEntityBatch {
+		go baseEntities.fetch(id, baseEntities.fields, baseEntities.version, true)
+	}
+	baseEntities.mu.Unlock()
+	iter.baseEntityBatch = iter.baseEntityBatch[:0]
+
+	for _, e := range iter.entityBatch {
+		if !iter.send(e) {
+			return false
+		}
+	}
+	iter.entityBatch = iter.entityBatch[:0]
+	return true
+}
+
+// send sends the given entity on iter.entityc.
+// It reports whether that entity was sent OK (that is,
+// the iterator has not been closed).
+func (iter *Iter) send(e *mongodoc.Entity) bool {
+	select {
+	case iter.entityc <- e:
+		return true
+	case <-iter.closed:
+		return false
+	}
+}
+
+func baseURL(url *charm.URL) *charm.URL {
```

Comment:
This function shows up all over the place, can we put it in one place and export it please?